File: go/simple/v2/ampli/ampli_test.go
package ampli

import (
	"testing"

	"github.com/amplitude/analytics-go/amplitude"
	"github.com/stretchr/testify/assert"
)

func TestIdentify(t *testing.T) {
	identify := Identify.Builder().
		SetRequiredNumber(6.4).
		SetOptionalArray([]string{"a", "b"}).Build()

	userProperties := map[amplitude.IdentityOp]map[string]interface{}{}
	userProperties[amplitude.IdentityOpSet] = map[string]interface{}{
		"requiredNumber": 6.4,
		"optionalArray":  []string{"a", "b"},
	}
	assert.Equal(t, userProperties, identify.toAmplitudeEvent().UserProperties)
}

func TestGroup(t *testing.T) {
	group := Group.Builder().SetRequiredBoolean(true).Build()

	expectEvent := amplitude.Event{
		EventType: amplitude.GroupIdentifyEventType,
		GroupProperties: map[amplitude.IdentityOp]map[string]interface{}{
			amplitude.IdentityOpSet: {"requiredBoolean": true},
		},
	}

	assert.Equal(t, expectEvent, group.toAmplitudeEvent())
}

func TestEventWithAllProperties(t *testing.T) {
	event := EventWithAllProperties.Builder().
		SetRequiredArray([]string{"abc", "test"}).
		SetRequiredBool(true).
		SetRequiredEnum(EventWithAllProperties.RequiredEnum.Enum1).
		SetRequiredInteger(3).
		SetRequiredNumber(16.4).
		SetRequiredString("str").
		SetOptionalString("optional-string").
		Build()

	eventProperties := map[string]interface{}{
		"requiredArray":   []string{"abc", "test"},
		"requiredBool":    true,
		"requiredEnum":    EventWithAllProperties.RequiredEnum.Enum1,
		"requiredInteger": 3,
		"requiredNumber":  16.4,
		"requiredString":  "str",
		"optionalString":  "optional-string",
	}
	assert.Equal(t, eventProperties, event.toAmplitudeEvent().EventProperties)
}

func TestAmpli_LoadWithEnvironment(t *testing.T) {
	instance := Ampli{}
	APIKey[EnvironmentDevelopment] = "test-development-api-key"
	instance.Load(LoadOptions{
		Environment: EnvironmentDevelopment,
	})

	assert.Equal(t, "test-development-api-key", instance.Client.Config().APIKey)
	assert.Equal(t, amplitude.DefaultConfig.FlushQueueSize, instance.Client.Config().FlushQueueSize)
}

func TestAmpli_LoadWithClient(t *testing.T) {
	instance := Ampli{}
	instance.Load(LoadOptions{
		Client: LoadClientOptions{
			APIKey: "test-development-api-key",
		},
	})

	assert.Equal(t, "test-development-api-key", instance.Client.Config().APIKey)
	assert.Equal(t, amplitude.DefaultConfig.FlushQueueSize, instance.Client.Config().FlushQueueSize)
}

func TestAmpli_LoadWithInstance(t *testing.T) {
	instance := Ampli{}
	instance.Load(LoadOptions{
		Client: LoadClientOptions{
			Instance: NewClient(NewClientConfig("test-development-api-key")),
		},
	})

	assert.Equal(t, "test-development-api-key", instance.Client.Config().APIKey)
	assert.Equal(t, amplitude.DefaultConfig.FlushQueueSize, instance.Client.Config().FlushQueueSize)
}

func TestAmpli_LoadWithConfig(t *testing.T) {
	instance := Ampli{}
	instance.Load(LoadOptions{
		Client: LoadClientOptions{
			Configuration: NewClientConfig("test-development-api-key"),
		},
	})

	assert.Equal(t, "test-development-api-key", instance.Client.Config().APIKey)
	assert.Equal(t, amplitude.DefaultConfig.FlushQueueSize, instance.Client.Config().FlushQueueSize)
}

File: browser/javascript/v2/ember-app/tests/unit/ampli-test.js
import { module, test } from 'qunit';
import { Ampli } from 'ember-app/ampli';

const send = () => ({
  status: 'success',
  statusCode: 200,
  body: {
    eventsIngested: 1,
    payloadSizeBytes: 1,
    serverUploadTime: 1,
  },
});

module('Unit | ampli', function () {
  test('should identify()', async function (assert) {
    const ampli = new Ampli();
    await ampli.load({
      client: {
        apiKey: 'test-api-key',
        configuration: {
          transportProvider: {
            send,
          },
        },
      },
    })?.promise;

    const response = await ampli.identify('test-user-id', {
      optionalArray: ['A', 'ray'],
      requiredNumber: 42,
    })?.promise;

    assert.strictEqual(response.event.event_type, '$identify');
    assert.strictEqual(response.event.user_id, 'test-user-id');
    assert.deepEqual(response.event.user_properties, {
      $set: {
        optionalArray: ['A', 'ray'],
        requiredNumber: 42,
      },
    });
  });

  test('should track an event with properties of all types', async function (assert) {
    const ampli = new Ampli();
    await ampli.load({
      client: {
        apiKey: 'test-api-key',
        configuration: {
          transportProvider: {
            send,
          },
        },
      },
    })?.promise;

    const response = await ampli.eventWithAllProperties({
      requiredBoolean: false,
      requiredInteger: 42,
      requiredEnum: 'Enum1',
      requiredNumber: 42.0,
      requiredString: 'Required string',
      requiredArray: ['Required', 'string'],
    })?.promise;

    assert.strictEqual(response.event.event_type, 'Event With All Properties');
    assert.deepEqual(response.event.event_properties, {
      requiredBoolean: false,
      requiredConst: 'some-const-value',
      requiredInteger: 42,
      requiredEnum: 'Enum1',
      requiredNumber: 42.0,
      requiredString: 'Required string',
      requiredArray: ['Required', 'string'],
    });
  });
});
